StonehearthEditor/Recipes/RecipesView.cs

```diff
@@ -18,64 +18,73 @@ public partial class RecipesView : UserControl
         public const string kAllCol = "All Columns";
 
         private bool mBaseModsOnly = true;
-
-        // Cell with extra getters for the DataRow/Column
-        private struct Cell
-        {
-            public RecipeRow Row { get; private set; }
-
-            public DataColumn Column { get; private set; }
-
-            public Cell(RecipeRow row, DataColumn column)
-            {
-                this.Row = row;
-                this.Column = column;
-            }
-        }
+        private bool mIsLoading = false;
+        private bool mIsApplyingChanges = false;
 
         private RecipeTable mDataTable;
-        private HashSet<Cell> mModifiedCells = new HashSet<Cell>();
-        private HashSet<int> mCmbxColumns = new HashSet<int>();
+        private HashSet<DataCell> mModifiedCells = new HashSet<DataCell>();
+        private HashSet<JsonFileData> mModifiedFiles = new HashSet<JsonFileData>();
+        private HashSet<int> mComboBoxColumns = new HashSet<int>();
+        private HashSet<ComboBox> mComboBoxes = new HashSet<ComboBox>();
+
+        private Stack<MultipleCellChange> mUndoStack = new Stack<MultipleCellChange>();
+        private Stack<MultipleCellChange> mRedoStack = new Stack<MultipleCellChange>();
+        private int currentStateUndoIndex = 0;
 
         // Cached material image paths
         private Dictionary<string, string> mMaterialImages = new Dictionary<string, string>();
 
         public RecipesView()
         {
             mDataTable = new RecipeTable(this);
-            mDataTable.ColumnChanged +=
-                (sender, e) =>
+            mDataTable.ColumnChanging +=
```

Comment:
It looks like this method is executed even if trying to set the value results in an error and no changes applied, so it's added to the stack and if you undo and redo, you get an uncaught exception.

Reply:
Can you give me an example? I haven't been able to reproduce this yet. Most changes I tried didn't trigger this event if it caused an error

Reply:
Ok should be fixed now